### PR TITLE
GH-117108: Change the size of the GC increment to about 1% of the total heap size.

### PR DESCRIPTION
**Pull request by markshannon**
The previous design of choosing increments based on the number of survivors doesn't work.
In order to handle the case of code that produces nothing but cycles, the rate of collection needs to be so high that the case of code that produces no cycles is slowed down hugely.

Fortunately, the fix is quite simple: base the size of the increment on the size of the heap.
Essentially we perform a whole-heap collection at the same rate as the traditional cycle collection, but in increments.
In the old collector, the thresholds were set so that a full heap collection was done every 100 young collections. We now do a 1% increment with every young collection.

* Issue: gh-117108

### CODE REVIEW

**Comment by mdboom** [comment]
Seems like this one is worth benchmarking (even though our gc benchmarks are a bit non-sensical, we may some things in the other ones).  I fired off a run.

**Review by mdboom — COMMENTED** [review]
I don't feel qualified to comment on the whole thing, but wanted to leave a reminder from #117123 to confirm it compiles with `--enable-pystats` and any new stats are added to `summarize_stats.py`.

**Comment by tim-one** [comment]
There are a couple compile-time warnings on the Windows build:
```
C:\Code\Python\Python\gc.c(1240,30): warning C4244: 'function': conversion from 'uintptr_t' to 'int', possibl
e loss of data [C:\Code\Python\PCbuild\pythoncore.vcxproj]
```
My guess is that `gc_set_old_space()` should be declared as having a fatter `size` argument than `int`.
```
C:\Code\Python\Python\gc.c(1700,54): warning C4244: 'function': conversion from 'Py_ssize_t' to 'int', possib
le loss of data [C:\Code\Python\PCbuild\pythoncore.vcxproj]
```
And here I guess `_PyGC_GetObjects()`'s `generation` argument should be `int` instead of `Py_ssize_t`.

**Comment by mdboom** [comment]
This PR overall improves speed by [27% GM / 0% HPT on the benchmarking suite and uses 6% more memory](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240321-3.13.0a5+-d0672a4), however, this is after the regression in #116206 of [24% GM / 0% HPT slower, using 1% less memory](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240319-3.13.0a5+-f4f04d6).  Running a custom comparison between the commit just before #116206 and this PR, it's an overall [improvement of 2% GM / 0% HPT](https://gist.github.com/mdboom/b2dd22a6a351db9ce9192bacf7dbd44a) for the GC changes + the bugfix in this PR.

In hindsight, we probably should have measured and not merged #116206 to begin with, because the regression is so obvious in some benchmarks.

(Here GM == Geometric mean, HPT == [Heirarchical Performance Testing](https://github.com/faster-cpython/benchmarking-public?tab=readme-ov-file#hpt) method).

**Comment by mdboom** [comment]
> > Given #117123, and the performance regression, maybe it makes sense to revert #116206 while working on getting this right?
> 
> Why not merge this PR? Your numbers show that it is faster than before #116206

That's fine, too.  I was just concerned about giving it the time it needs for a review (seeing none).

File: Lib/test/test_gc.py
Comment (ericsnowcurrently) [inline Lib/test/test_gc.py:1069]: Won't this leave GC disabled when the last test in this class is done?  Subsequent tests would then run with GC disabled, right?  If that's true, then perhaps there should be a `tearDownClass()` that re-enables GC one last time.

File: Lib/test/test_gc.py
Comment (ericsnowcurrently) [inline Lib/test/test_gc.py]:
```suggestion
        # There will be some objects we aren't counting,
```

File: Lib/test/test_gc.py
Comment (ericsnowcurrently) [inline Lib/test/test_gc.py:1103]: What's the likelihood that this number might need to be changed in the future?  Keep in mind that readers that aren't as familiar with GC (especially the incremental version) will be hard-pressed to know which objects aren't counted (nor how many there might be).  Any additional information you can leave in this note would be supremely helpful.

File: Python/gc.c
Comment (ericsnowcurrently) [inline Python/gc.c:1317]: IIUC, this is the core of the fix.  Is that right?

And just to be sure I understood, this has the concrete effect of making GC collection happen more frequently.  Is that right?

Without having more than cursory knowledge about the GC implementation, the change in frequency seems like a significant change relative to the performance profile of GC (which is the point of the fix, of course).  The relative effect is clearly derived from the overhead of collection vs. the cost of larger collections (and the continual cost of larger generations).  Are there other tradeoffs to which we need to pay attention?

File: Python/gc.c
Comment (ericsnowcurrently) [inline Python/gc.c]: FWIW, `gcstate->old[gcstate->visited_space^1]` is about as unclear to me it could possibly be, as a reader with very little context about the GC implementation.  I'm not saying you need to do anything about this, but just wanted to point out my experience, which I expect most people would share.  (I do suppose that the GC implementation is saturated with tricky code that requires sufficient fore-knowledge to follow.)

Is there any single document that condenses all the context into one spot?  If not, it might be worth adding one, or at least adding comments where necessary to give at least some context for the trickier spots.

File: Python/gc.c
Comment (ericsnowcurrently) [inline Python/gc.c:1431]: That's a nice little simplification. 😄

IIUC, instead of calling `gc_list_size()` with every increment, we increment/decrement a `PyInterpreterState` field with the link/del of each GC object.  Is that right?

Is it part of the fix or more of a drive-by improvement?  I suppose I can see how this might result in better performance, especially with smaller collections.

File: Python/gc.c
Comment (ericsnowcurrently) [inline Python/gc.c:1414]: FWIW, I'm a fan of keeping declarations close to where they are used, as you have here. 👏

**Review by ericsnowcurrently — APPROVED** [review]
First of all, I want to be clear that I have only a basic understanding of the GC implementation, and perhaps even less so of the new incremental GC.  Consequently, the confidence interval around my assessment here is quite large.

Even from that point of view, the fix in this PR seems straightforward and the correlation to an impact on performances is clear.  However, if you think this PR might have any subtle secondary effects to consider then you should get an additional review from someone more expert.

I'm approving the PR under the assumption that you will address my few concerns and answer my questions (and get an expert review, if necessary) before merging.

File: Lib/test/test_gc.py
Comment (markshannon) [reply Lib/test/test_gc.py:1069]: This module disables gc [during setup](https://github.com/python/cpython/blob/main/Lib/test/test_gc.py#L1481) and re-enables it at the end, so this (and the `setUp` method) just turns GC back on this for this test.

File: Lib/test/test_gc.py
Comment (markshannon) [reply Lib/test/test_gc.py:1103]: As it says, it is just a estimate. 
I added this in an attempt to make the test stricter, so it is more likely to catch a regression, than anything exact.

File: Python/gc.c
Comment (markshannon) [reply Python/gc.c:1317]: Yes, this is the important change.

It may make GC happen more frequently, but is more likely to make it happen *less* frequently.
The important thing is that the long term rate of incremental GC is the same as long term rate of the old generation collection prior to https://github.com/python/cpython/pull/116206.

(Young generation collection rates are unchanged)

File: Python/gc.c
Comment (markshannon) [reply Python/gc.c]: I changed this to avoid the unused variable error.
I've changed it back and added an `#ifdef` around it.

File: Lib/test/test_gc.py
Comment (gpshead) [reply Lib/test/test_gc.py:1069]: I suggest leaving a comment to that effect.  _(we often see code in huge files like this in bits and pieces so would otherwise not realize this in things like this diff review context)_

File: Python/gc.c
Comment (markshannon) [reply Python/gc.c:1431]: This is the key part of the fix (along with the change in the young collection above).

Basing the amount of work to do on the survivor count doesn't seem to work.
Basing it on the total heap size is known to work.

File: Modules/gcmodule.c
Comment (gpshead) [inline Modules/gcmodule.c:329]: why not just change the type to int in the clinic input above?

**Comment by mdboom** [comment]
Looking deeper at the benchmarking data for this PR vs. the commit prior to #116206, memory usage is up pretty significantly for the benchmarks that actually have a lot of data.  (By 45% for mypy2, for example).  Maybe that doesn't matter as long as the memory usage still has a reasonable upper bound, but it's something to probably confirm.

<details>
<summary>Memory usage of this PR vs. prior to incremental GC</summary>

image
</details>

**Comment by markshannon** [comment]
OOI, how much in absolute terms? A relative increase in memory isn't that meaningful.

We expect short lived programs to have increased memory use as the collector will lag a bit more than before, being an incremental collector.

File: Modules/gcmodule.c
Comment (markshannon) [reply Modules/gcmodule.c:329]: It gives a slightly better error message for values that don't fit in an `int`.
Since it's vanishingly unlikely for anyone to pass in that large a number, I doubt it matters.

Currently if you pass in `1 << 32` you get "generation parameter must be less than the number of  available generations", instead of the argument clinic overflow message.

**Comment by mdboom** [comment]
> OOI, how much in absolute terms? A relative increase in memory isn't that meaningful.

Here's the absolute numbers.

I think what really matters for this is to ensure that highly cyclical data structures are still collected in a reasonable time frame.  I think the only way to get that from the data would be to increase the number of iterations and see how that affects memory usage.

<details>
<summary>Absolute differences in memory usage</summary

| benchmark | base | head | diff |
| --- | ---: | ---: | ---: |
| 2to3 | 23,279,323 | 23,574,625 | 295,302 |
| aiohttp | 30,047,865 | 29,992,862 | -55,003 |
| async_generators | 36,232,825 | 36,214,101 | -18,724 |
| async_tree_none | 80,446,220 | 80,228,937 | -217,283 |
| async_tree_cpu_io_mixed | 83,602,090 | 83,824,640 | 222,550 |
| async_tree_cpu_io_mixed_tg | 92,690,529 | 93,186,925 | 496,396 |
| async_tree_io | 136,706,145 | 135,888,700 | -817,445 |
| async_tree_io_tg | 145,445,644 | 146,071,747 | 626,103 |
| async_tree_memoization | 86,078,610 | 85,938,371 | -140,239 |
| async_tree_memoization_tg | 95,327,768 | 95,374,189 | 46,421 |
| async_tree_none_tg | 89,782,369 | 89,443,961 | -338,408 |
| asyncio_tcp | 50,583,454 | 51,326,390 | 742,936 |
| asyncio_tcp_ssl | 76,057,648 | 76,106,020 | 48,372 |
| asyncio_websockets | 65,070,811 | 64,648,728 | -422,083 |
| chameleon | 29,368,905 | 29,676,300 | 307,395 |
| chaos | 17,821,110 | 17,829,107 | 7,997 |
| comprehensions | 19,322,782 | 19,364,717 | 41,935 |
| bench_mp_pool | 20,654,762 | 20,830,890 | 176,128 |
| coroutines | 17,117,379 | 17,019,270 | -98,109 |
| crypto_pyaes | 17,455,981 | 17,514,691 | 58,710 |
| dask | 88,305,273 | 88,550,448 | 245,175 |
| deepcopy | 18,866,956 | 18,866,176 | -780 |
| deepcopy_reduce | 18,834,578 | 18,855,643 | 21,065 |
| deepcopy_memo | 18,891,727 | 18,836,918 | -54,809 |
| deltablue | 19,162,648 | 21,385,216 | 2,222,568 |
| django_template | 43,785,069 | 43,736,697 | -48,372 |
| djangocms | 30,294,016 | 30,260,467 | -33,549 |
| docutils | 64,943,835 | 85,793,840 | 20,850,005 |
| dulwich_log | 26,350,348 | 25,903,299 | -447,049 |
| fannkuch | 17,007,957 | 16,951,393 | -56,564 |
| float | 34,400,158 | 34,406,009 | 5,851 |
| create_gc_cycles | 17,081,685 | 17,093,973 | 12,288 |
| gc_traversal | 20,627,651 | 20,528,566 | -99,085 |
| generators | 29,539,571 | 29,482,227 | -57,344 |
| genshi_text | 20,681,094 | 21,016,576 | 335,482 |
| genshi_xml | 21,943,832 | 22,484,114 | 540,282 |
| go | 20,013,641 | 18,889,191 | -1,124,450 |
| gunicorn | 30,099,553 | 29,992,667 | -106,886 |
| hexiom | 18,173,561 | 18,145,475 | -28,086 |
| html5lib | 37,427,687 | 41,330,590 | 3,902,903 |
| json | 18,928,006 | 18,950,241 | 22,235 |
| json_dumps | 17,250,986 | 17,288,630 | 37,644 |
| json_loads | 17,275,172 | 17,305,600 | 30,428 |
| logging_format | 24,717,994 | 24,727,942 | 9,948 |
| logging_silent | 17,715,395 | 17,760,256 | 44,861 |
| logging_simple | 23,280,688 | 23,272,691 | -7,997 |
| mako | 26,486,881 | 26,608,396 | 121,515 |
| mdp | 25,583,225 | 25,669,827 | 86,602 |
| meteor_contest | 19,434,739 | 19,383,442 | -51,297 |
| mypy2 | 167,246,116 | 242,955,605 | 75,709,489 |
| nbody | 17,175,308 | 17,166,336 | -8,972 |
| nqueens | 17,238,308 | 17,207,491 | -30,817 |
| pathlib | 18,659,035 | 18,834,383 | 175,348 |
| pickle | 17,929,362 | 18,064,140 | 134,778 |
| pickle_dict | 17,877,869 | 17,895,228 | 17,359 |
| pickle_list | 17,998,799 | 17,963,885 | -34,914 |
| pickle_pure_python | 17,842,371 | 17,920,975 | 78,604 |
| pidigits | 17,161,264 | 17,136,883 | -24,381 |
| pprint_safe_repr | 37,872,591 | 37,859,718 | -12,873 |
| pprint_pformat | 37,840,798 | 37,846,454 | 5,656 |
| pycparser | 39,668,784 | 46,512,420 | 6,843,636 |
| pyflate | 60,085,199 | 59,582,366 | -502,833 |
| pylint | 92,368,310 | 95,753,752 | 3,385,442 |
| python_startup | 13,590,332 | 13,637,729 | 47,397 |
| python_startup_no_site | 13,584,871 | 13,620,175 | 35,304 |
| raytrace | 17,892,108 | 17,909,662 | 17,554 |
| regex_dna | 25,022,854 | 25,080,393 | 57,539 |
| regex_effbot | 17,865,971 | 17,899,520 | 33,549 |
| regex_v8 | 23,517,086 | 23,460,327 | -56,759 |
| richards | 17,694,720 | 17,702,912 | 8,192 |
| richards_super | 17,723,392 | 17,667,413 | -55,979 |
| scimark_fft | 18,076,428 | 18,055,168 | -21,260 |
| scimark_lu | 17,692,769 | 17,755,769 | 63,000 |
| scimark_monte_carlo | 18,058,873 | 18,048,926 | -9,947 |
| scimark_sparse_mat_mult | 18,079,353 | 18,216,667 | 137,314 |
| spectral_norm | 16,931,693 | 17,008,932 | 77,239 |
| sqlglot_normalize | 22,803,407 | 22,787,218 | -16,189 |
| sqlglot_optimize | 23,573,845 | 23,881,045 | 307,200 |
| sqlglot_parse | 22,958,665 | 25,704,350 | 2,745,685 |
| sqlglot_transpile | 23,045,461 | 24,079,993 | 1,034,532 |
| sqlite_synth | 19,792,457 | 19,782,509 | -9,948 |
| sympy_expand | 59,473,139 | 59,392,975 | -80,164 |
| sympy_integrate | 60,309,504 | 60,277,126 | -32,378 |
| sympy_sum | 72,773,632 | 72,791,771 | 18,139 |
| sympy_str | 60,266,203 | 60,242,602 | -23,601 |
| telco | 17,062,180 | 17,114,453 | 52,273 |
| thrift | 18,817,609 | 18,836,528 | 18,919 |
| tomli_loads | 180,438,162 | 177,827,254 | -2,610,908 |
| tornado_http | 41,430,649 | 41,497,356 | 66,707 |
| typing_runtime_protocols | 19,174,156 | 19,157,772 | -16,384 |
| unpack_sequence | 18,567,753 | 18,615,929 | 48,176 |
| unpickle | 18,123,239 | 18,082,474 | -40,765 |
| unpickle_list | 18,007,186 | 18,036,248 | 29,062 |
| unpickle_pure_python | 17,914,343 | 17,938,529 | 24,186 |
| xml_etree_parse | 22,025,362 | 22,069,443 | 44,081 |
| xml_etree_iterparse | 22,826,227 | 22,949,107 | 122,880 |
| xml_etree_generate | 23,017,569 | 23,610,514 | 592,945 |
| xml_etree_process | 22,720,121 | 22,722,462 | 2,341 |
</details>

**Comment by markshannon** [comment]
I'm going to merge this, as it does fix the immediate problem.

The memory use for mypy2 is up significantly, but the memory use for "tomli_loads", which also uses a significant amount of memory, is down.
I'll leave the issue open and we can look to fine tune thresholds later.

**Comment by mdboom** [comment]
> I'm going to merge this, as it does fix the immediate problem.

Ok, and you are going to fix the pystats build as a follow-on, rather than addressing here?

**Comment by encukou** [comment]
Since this PR was merged, many buildbots started consistently failing in `test_multiprocessing` tests. For example: https://buildbot.python.org/all/#/builders/725/builds/7475

I'm not sure this can be fixed quickly. If it can't, we'll need to revert the feature.

**Comment by markshannon** [comment]
https://github.com/python/cpython/pull/117213 should fix it

**Comment by vstinner** [comment]
> Since this PR was merged, many buildbots started consistently failing in test_multiprocessing tests. For example: https://buildbot.python.org/all/#/builders/725/builds/7475

FYI I created PR gh-117252 to track this failure to help triaging buildbot failures.

**Pull request merged**
